pkg/controller/machine_test.go

```diff
@@ -290,4 +278,276 @@ var _ = Describe("machine", func() {
 			}, machinev1.MachineRunning),
 		)
 	})
+
+	Describe("#ValidateMachine", func() {
+		type data struct {
+			action machineapi.Machine
+			expect field.ErrorList
+		}
+		DescribeTable("#happy path",
+			func(data *data) {
+				errList := validation.ValidateMachine(&data.action)
+				Expect(errList).To(Equal(data.expect))
+			},
+			Entry("aws", &data{
+				action: machineapi.Machine{
+					Spec: machineapi.MachineSpec{
+						Class: machineapi.ClassSpec{
+							Kind: "AWSMachineClass",
+							Name: "aws",
+						},
+					},
+				},
+				expect: field.ErrorList{},
+			}),
+		)
+	})
+
+	Describe("#validateMachineClass", func() {
+		type setup struct {
+			aws     []*machinev1.AWSMachineClass
+			secrets []*corev1.Secret
+		}
+		type expect struct {
+			machineClass interface{}
+			secret       *corev1.Secret
+			err          bool
+		}
+		type data struct {
+			setup  setup
+			action *machinev1.ClassSpec
+			expect expect
+		}
+
+		objMeta := &metav1.ObjectMeta{
+			GenerateName: "class",
+			Namespace:    namespace,
+		}
+
+		DescribeTable("##table",
```

Comment:
Seems to be. But I am not that particular. Ginkgo has a good detailed output format anyway.

Reply:
Okay thanks 👍

Reply:
Well, Describe and Context in Ginkgo are container blocks to organise your test specs and they can be nested - you have define one or more Describe container block within a top level Describe container block. It helps to give meaningful texts - improves the readability, maintainability. Also, in case you want to generate reports it helps. You can check some examples to get an idea. I have tried to follow that pattern in my test cases